src/main/java/com/squareup/kotlinpoet/TypeVariableName.kt

```diff
@@ -65,32 +66,52 @@ class TypeVariableName private constructor(
   }
 
   companion object {
-    internal fun of(name: String, bounds: List<TypeName>): TypeVariableName {
+    internal fun of(name: String, bounds: List<TypeName>, variance: KModifier?): TypeVariableName {
+      require(variance == null || variance == KModifier.IN || variance == KModifier.OUT) {
+        "$variance is an invalid variance modifier, the only allowed values are in and out!"
+      }
       // Strip java.lang.Object from bounds if it is present.
-      return TypeVariableName(name, bounds.filter { it != ANY })
+      return TypeVariableName(name, bounds.filter { it != ANY }, variance)
     }
 
     /** Returns type variable named `name` without bounds.  */
     @JvmStatic @JvmName("get")
-    operator fun invoke(name: String) = TypeVariableName.of(name, emptyList())
+    operator fun invoke(name: String) = TypeVariableName.of(name, emptyList(), variance = null)
+
+    /** Returns type variable named `name` with `variance` and without bounds.  */
+    @JvmStatic @JvmName("get")
+    operator fun invoke(name: String, variance: KModifier) =
+        TypeVariableName.of(name, emptyList(), variance)
 
     /** Returns type variable named `name` with `bounds`.  */
     @JvmStatic @JvmName("get")
-    operator fun invoke(name: String, vararg bounds: TypeName): TypeVariableName {
-      return TypeVariableName.of(name, bounds.toList())
-    }
+    operator fun invoke(name: String, vararg bounds: TypeName) =
+        TypeVariableName.of(name, bounds.toList(), variance = null)
+
+    /** Returns type variable named `name` with `variance` and `bounds`.  */
+    @JvmStatic @JvmName("get")
+    operator fun invoke(name: String, variance: KModifier, vararg bounds: TypeName) =
+        TypeVariableName.of(name, bounds.toList(), variance)
 
     /** Returns type variable named `name` with `bounds`.  */
     @JvmStatic @JvmName("get")
-    operator fun invoke(name: String, vararg bounds: KClass<*>): TypeVariableName {
-      return TypeVariableName.of(name, bounds.map { it.asTypeName() })
-    }
+    operator fun invoke(name: String, vararg bounds: KClass<*>) =
+        TypeVariableName.of(name, bounds.map { it.asTypeName() }, variance = null)
+
+    /** Returns type variable named `name` with `variance` and `bounds`.  */
+    @JvmStatic @JvmName("get")
+    operator fun invoke(name: String, variance: KModifier, vararg bounds: KClass<*>) =
```

Comment:
I tinkered with the factory methods a bit and pushed an alternative version. Now, to pass both `bounds` and `variance` in Kotlin, you have to use named arguments:

```kotlin
TypeVariableName("E", Number::class, variance = KModifier.IN)
```
In my opinion, that's not ideal, but since `variance` is an optional parameter, I think it's fine. Additionally, I generally like using named arguments, as it improves readability in most cases.

If you want to do the same in Java, you'll have to pass an array instead of varargs:

```java
TypeVariableName.get("name", new TypeName[]{null}, KModifier.OUT);
```
Given that we're probably not aiming at perfectly Java-friendly API, I think it's alright. What do you think?